feat: Add support for project titles

marc1404 opened the pull request:
**What this PR does / why we need it**:

This PR adds support for `Project` "display names" via a new annotation:
`dashboard.gardener.cloud/project-title: "..."`

See the issue for more context:
* #2469 

**Screenshots:**

_Navigation on the left-hand side:_
<img width="245" height="446" alt="Screenshot 2025-08-01 at 15 56 35" src="https://github.com/user-attachments/assets/2fb7e7e8-bea6-4be3-991b-33be98399c33" />

_Expanded navigation on the left-hand side:_
<img width="248" height="568" alt="Screenshot 2025-08-01 at 15 56 50" src="https://github.com/user-attachments/assets/4787816c-1020-42a8-9cc4-0a115910bc70" />

_`Shoot` cluster list in the "All Projects"-scope:_
<img width="1690" height="435" alt="Screenshot 2025-08-01 at 15 56 59" src="https://github.com/user-attachments/assets/e8e12a15-b9fa-4989-bd67-0930078571b7" />

_Project creation dialog:_
<img width="595" height="486" alt="Screenshot 2025-08-01 at 15 57 36" src="https://github.com/user-attachments/assets/cb6e4796-b983-4c06-86da-5cb6fff82fee" />

_Project administration view:_
<img width="1221" height="725" alt="Screenshot 2025-08-01 at 15 57 15" src="https://github.com/user-attachments/assets/4ad9c192-8083-4ffe-bbc8-cc2c15b68bb2" />

**Which issue(s) this PR fixes**:

Fixes #2469

**Special notes for your reviewer**:

/cc @klocke-io @benedikt-haug 
/cc @petersutter @grolu 

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```feature user
Added support for `Project` titles through a new annotation `dashboard.gardener.cloud/project-title`.
```

marc1404 commented:
/hold

🚧 _work in progress..._

klocke-io commented:
/ok-to-test

grolu commented:
Could you add some screenshots

grolu commented:
>Isn't there some chance to strike a middle ground here?

Sure, as I already pointed out in the kickoff call, this is the main challenge :)
Implementing a project title from a technical perspective is simple. But integrating it in a way that truly adds value—while still ensuring the user always knows which project they are looking at—is the key point.
I'm open to suggestions!

marc1404 commented on frontend/src/components/editable/GEditableText.vue:
```diff
@@ -67,6 +67,8 @@ SPDX-License-Identifier: Apache-2.0
         :messages="messages"
         :error-messages="getErrorMessages(v$.internalValue)"
         class="g-field"
+        :counter="counter"
+        :maxlength="maxLength"
```
Thanks! Fixed via: [`2acc5b7` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/2acc5b7fbcf4fdec46cc0bbe2d696cec97c04ef4)

marc1404 commented on frontend/src/components/dialogs/GProjectDialog.vue:
```diff
@@ -27,7 +27,31 @@ SPDX-License-Identifier: Apache-2.0
+                location="top"
+                text="Human-readable project title."
+              />
+            </v-text-field>
```
Fixed via: [`7a83d67` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/7a83d676b603b15a6d563d8f0c51054b28fa5486)

marc1404 commented on frontend/src/views/GAdministration.vue:
```diff
@@ -39,6 +44,32 @@ SPDX-License-Identifier: Apache-2.0
+                  </template>
+                  <div class="text-body-2 text-medium-emphasis">
+                    <span>Title</span>
+                    <v-tooltip
```
Fixed via: [`164751c` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/164751c432c7a892724f70659bbe9ecec61ea46c)

marc1404 commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -129,10 +135,37 @@ SPDX-License-Identifier: Apache-2.0
-                {{ getProjectOwner(project) }}
+                <abbr :title="hasProjectTitle(project) ? '' : null">
+                  <span v-if="hasProjectTitle(project)">
+                    {{ project.metadata.name }} &bull;
```
Fixed via: [`bb0bc63` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/bb0bc63da4ea3b399b4e1ca63bb71d44af28049a)

marc1404 commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -348,6 +383,10 @@ const selectedProjectName = computed(() => {
+const selectedProjectTitle = computed(() => {
+  return getProjectTitle(selectedProject.value) ?? selectedProjectName
+})
+
```
Fixed via: [`bb0bc63` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/bb0bc63da4ea3b399b4e1ca63bb71d44af28049a#diff-43afdd385b0352b8183bb039a846f90d360a5ce4dce53e76c51f963599611c18R351)

marc1404 commented on frontend/src/views/GAdministration.vue:
```diff
@@ -522,6 +553,8 @@ import {
   getDateFormatted,
 } from '@/utils'
 import { errorDetailsFromError } from '@/utils/error'
+import { annotations } from '@/utils/annotations.js'
```
Fixed via: [`c2f5dae` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/c2f5dae6f4d3a721daaeb21e9310350e85d0c581)

marc1404 commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -131,6 +143,9 @@ SPDX-License-Identifier: Apache-2.0
               <v-list-item-title class="project-name text-uppercase">
                 {{ project.metadata.name }}
               </v-list-item-title>
+              <v-list-item-title class="project-title">
+                {{ getProjectTitle(project) }}
+              </v-list-item-title>
```
A similar change is made to the project list as above. The fixed height of `48px` is removed to allow for vertically larger items when the project title is set. The project title uses the same font size as the owner and uses normal casing to be less emphasized than the technical project name.
<img width="250" height="486" alt="Image" src="https://github.com/user-attachments/assets/759a3b94-8826-4297-a709-8bd4a8de61bf" />

marc1404 commented on frontend/src/views/GAdministration.vue:
```diff
@@ -461,7 +496,7 @@ SPDX-License-Identifier: Apache-2.0
       </template>
       <template #content>
         <v-card-text>
-          Are you sure to delete the project <span class="font-weight-bold">{{ projectName }}</span>?
+          Are you sure to delete the project <span class="font-weight-bold">{{ projectNameAndTitle }}</span>?
```
<img width="703" height="267" alt="Image" src="https://github.com/user-attachments/assets/add6fa34-d263-4e0d-bd0a-dcc1981ae95f" />

marc1404 commented on frontend/src/views/GMembers.vue:
<img width="605" height="269" alt="Image" src="https://github.com/user-attachments/assets/5966ae3a-a3b7-44ba-8f6a-879ad43b7b38" />

marc1404 commented on frontend/src/App.vue:
```diff
@@ -85,12 +85,12 @@ const documentTitle = computed(() => {
+    const projectNameAndTitle = projectStore.projectNameAndTitle
     const routeParamName = route.params.name
     if (routeParamName) {
-      titleItems.push([projectName, routeParamName].join('/'))
-    } else if (projectName) {
-      titleItems.push(projectName)
+      titleItems.push([projectNameAndTitle, routeParamName].join('/'))
+    } else if (projectNameAndTitle) {
+      titleItems.push(projectNameAndTitle)
```
<img width="413" height="125" alt="Image" src="https://github.com/user-attachments/assets/6898e900-c7b5-4c76-971c-1b31cb09dc7a" />

marc1404 commented:
ℹ️ @klocke-io @grolu @petersutter the PR is ready for review again whenever you have time :)
I've changed:
* added a `GProjectTooltip` (similar to the `GVendor` component) that is used in all places where the `Project` title could be truncated [[`11ba11a` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/11ba11ac45250ec45e6ef61cefe6cfecef921295)].

Fields are shown conditionally when they're set:
<img width="589" height="623" alt="image" src="https://github.com/user-attachments/assets/bacda4db-9d29-485f-8ca7-f3109ca9194e" />

* how a `Project` title is displayed in the cluster list [[`e7464f6` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/e7464f656c76c401a941e6405fae6d47fb6f5860)].

<img width="615" height="609" alt="image" src="https://github.com/user-attachments/assets/f20d1445-cdab-4f4f-aa33-bd210d80b7fe" />

grolu commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -128,12 +142,21 @@ SPDX-License-Identifier: Apache-2.0
-                {{ getProjectOwner(project) }}
-              </v-list-item-subtitle>
+              <g-project-tooltip
+                open-delay="1000"
```
```suggestion
                :open-delay="1000"
```

grolu commented on frontend/src/composables/useProjectItem.js:
```diff
@@ -11,7 +11,7 @@ import {
   provide,
 } from 'vue'
 
-import { useProjectMetadata } from '@/composables/useProjectMetadata'
```
why did you add index.js to all the `useProjectMetadata` imports?

grolu commented on frontend/src/composables/useProjectItem.js:
```diff
@@ -11,7 +11,7 @@ import {
 } from 'vue'
 
-import { useProjectMetadata } from '@/composables/useProjectMetadata'
+import { useProjectMetadata } from '@/composables/useProjectMetadata/index.js'
```
```suggestion
import { useProjectMetadata } from '@/composables/useProjectMetadata'
```

grolu commented on frontend/src/components/GProjectTooltip.vue:
```diff
@@ -0,0 +1,91 @@
+  <v-tooltip
+    location="top"
+    :open-delay="openDelay"
+    :open-on-hover="openOnHover"
```
Prevent tooltips from growing too large. Descriptions may get truncated but I think we can live with that. Maybe we will redesign this popover at some point to fit more content

grolu commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -663,6 +720,10 @@ watch(projectMenu, value => {
         opacity: 0.9;
       }
 
+      :deep(.v-list-item__prepend > .v-list-item__spacer) {
+        width: 16px;
+      }
```
You could also get rid of the checkmark column and move all content more to the left, this would allow us to show even longer titles. We could indicate the selected project by using a (primary) highlight color or emphasize by sing bold text etc.

petersutter commented on frontend/src/composables/useProjectMetadata/index.js:
```diff
@@ -26,9 +30,34 @@ export const useProjectMetadata = (projectItem, options = {}) => {
     unsetLabel: unsetProjectLabel,
   } = useObjectMetadata(projectItem)
 
+  // projectTitle is a project-specific annotation
```
```suggestion
```

marc1404 commented on frontend/src/components/GMainNavigation.vue:
```diff
@@ -55,17 +55,32 @@ SPDX-License-Identifier: Apache-2.0
+                  </div>
+                </g-project-tooltip>
+
+                <div v-if="selectedProject">
```
I could not figure out a reason for changing it, therefore it's back to a `template` like it should be, thanks :)

marc1404 replied:
Fixed via: [`e7110b9` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/e7110b9128ea08159792c55ccf6e4b819864a364)

marc1404 commented on frontend/src/components/GProjectTooltip.vue:
```diff
@@ -0,0 +1,92 @@
+import { useProjectMetadata } from '@/composables/useProjectMetadata'
+import { useProvideProjectItem } from '@/composables/useProjectItem.js'
+
+const props = defineProps({
+  project: { type: Object, required: true },
+  openDelay: { type: Number, default: 200 },
+  openOnHover: { type: Boolean, default: true },
+})
+const { project, openDelay, openOnHover } = toRefs(props)
+const { projectName, projectTitle } = useProjectMetadata(project)
+const { projectDescription, projectOwner, projectCreatedBy, projectCreationTimestamp, projectPurpose } = useProvideProjectItem(project)
```
I'll do so :) Ideally, should this be automatically handled by ESLint/Prettier?

marc1404 commented on frontend/src/store/project.js:
```diff
@@ -93,6 +107,17 @@ export const useProjectStore = defineStore('project', () => {
+    return title ? title.trim() : title
+  })
+
+  const projectNameAndTitle = computed(() => {
+    const name = projectName.value
+    const title = projectTitle.value
+    return title ? `${name} – ${title}` : name
+  })
```
I never considered that the standard keyboard types an en dash 🙃 Eagle-eye! 🦅

marc1404 replied:
Fixed via:
* Reusable project name and title: [`b320b2f` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/b320b2fc9c95652d4dd2ebe0143230102e733f85)
* en dash -> em dash: [`9412855` (#2470)](https://github.com/gardener/dashboard/pull/2470/commits/941285550faaa8379f63fa26363ff5bdff4c85d8)
